src/ndt/protocol_impl.hpp

```diff
@@ -0,0 +1,260 @@
+        if (err) {
+            callback(ReadingSrvQueueMessageError(err));
+            return;
+        }
+        if (type != SRV_QUEUE) {
+            callback(NotSrvQueueMessageError());
+            return;
+        }
+        ErrorOr<unsigned> wait_time = lexical_cast_noexcept<unsigned>(s);
+        if (!wait_time) {
+            callback(InvalidSrvQueueMessageError(wait_time.as_error()));
+            return;
+        }
+        ctx->logger->info("Wait time before test starts: %d", *wait_time);
+        // XXX Simplified implementation
+        if (*wait_time > 0) {
+            callback(UnhandledSrvQueueMessageError());
+            return;
+        }
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(messages, read_msg)>
+void recv_version_impl(Var<Context> ctx, Callback<Error> callback) {
+    ctx->logger->debug("ndt: recv server version ...");
+    messages_read_msg(ctx, [=](Error err, uint8_t type, std::string s) {
+        ctx->logger->debug("ndt: recv server version ... %d", (int)err);
+        if (err) {
+            callback(ReadingServerVersionMessageError(err));
+            return;
+        }
+        if (type != MSG_LOGIN) {
+            callback(NotServerVersionMessageError());
+            return;
+        }
+        ctx->logger->info("Got server version: %s", s.c_str());
+        // TODO: validate the server version?
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(messages, read_msg)>
+void recv_tests_id_impl(Var<Context> ctx, Callback<Error> callback) {
+    ctx->logger->debug("ndt: recv tests ID ...");
+    messages_read_msg(ctx, [=](Error err, uint8_t type, std::string s) {
+        ctx->logger->debug("ndt: recv tests ID ... %d", (int)err);
+        if (err) {
+            callback(ReadingTestsIdMessageError(err));
+            return;
+        }
+        if (type != MSG_LOGIN) {
+            callback(NotTestsIdMessageError());
+            return;
+        }
+        ctx->logger->info("Authorized tests: %s", s.c_str());
+        ctx->granted_suite = split(s);
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(test_c2s, run), MK_MOCK_NAMESPACE(test_meta, run),
+          MK_MOCK_NAMESPACE(test_s2c, run)>
+void run_tests_impl(Var<Context> ctx, Callback<Error> callback) {
+
+    if (ctx->granted_suite.size() <= 0) {
+        callback(NoError());
+        return;
+    }
+
+    std::string s = ctx->granted_suite.front();
+    ctx->granted_suite.pop_front();
+
+    ErrorOr<int> num = lexical_cast_noexcept<int>(s);
+    if (!num) {
+        callback(InvalidTestIdError(num.as_error()));
+        return;
+    }
+
+    std::function<void(Var<Context>, Callback<Error>)> func;
+    switch (*num) {
+    case TEST_C2S:
+        func = test_c2s_run;
+        break;
+    case TEST_META:
+        func = test_meta_run;
+        break;
+    case TEST_S2C:
+        func = test_s2c_run;
+        break;
+    }
```

Comment:
Maybe it's good to add checks for all the supported NDT tests `TEST_MID`, `TEST_SFW ` (even if they are not yet implemented in NDT) and as suggested by the NDT design doc:

> The Client MUST drop the connection when it receives unknown test id, because such situation means an error on the Server's part